// @dialectlabs/blinks/dist/hooks/evm/index.js
// src/hooks/evm/useEvmWagmiAdapter.ts
import "@dialectlabs/blinks-core";
import { useMemo } from "react";
import { deserialize, useAccount, useChains, useConnectorClient } from "wagmi";
import {
  TransactionExecutionError
} from "viem";
import {
  sendTransaction,
  signMessage,
  waitForTransactionReceipt
} from "viem/actions";
function useEvmWagmiAdapter({
  onConnectWalletRequest
}) {
  const chains = useChains();
  const { address } = useAccount();
  const { data: client } = useConnectorClient();
  const adapter = useMemo(
    () => ({
      metadata: {
        supportedBlockchainIds: chains.map((chain) => `eip155:${chain.id}`)
      },
      connect: async () => {
        if (!address) {
          try {
            const potentialAddress = await onConnectWalletRequest();
            return potentialAddress || null;
          } catch (error) {
            console.error("Blink wagmi adapter connection failed:", error);
            return null;
          }
        }
        return address;
      },
      signTransaction: async (tx) => {
        if (!address || !client) {
          return { error: "No wallet connected" };
        }
        try {
          const deserializedTx = deserialize(tx);
          const hash = await sendTransaction(client, {
            ...deserializedTx,
            account: address
          });
          return { signature: hash };
        } catch (e) {
          console.error("Blink wagmi adapter sign transaction failed:", e);
          if (e instanceof TransactionExecutionError) {
            return { error: `Unable to sign: ${e.shortMessage}` };
          }
          return { error: "Signing failed" };
        }
      },
      signMessage: async (data) => {
        if (!address || !client) {
          return { error: "No wallet connected" };
        }
        let message;
        if (typeof data === "string") {
          message = data;
        } else if (typeof data === "object" && "raw" in data) {
          message = data;
        } else {
          message = JSON.stringify(data);
        }
        try {
          const signature = await signMessage(client, {
            account: address,
            message
          });
          return { signature };
        } catch (e) {
          console.error("Blink wagmi adapter message signing failed:", e);
          return { error: "Message signing failed" };
        }
      },
      confirmTransaction: async (signature) => {
        if (!client) {
          throw new Error("No wallet connected");
        }
        const res = await waitForTransactionReceipt(client, {
          hash: signature
        });
        if (res.status !== "success") {
          throw new Error("Transaction confirmation failed");
        }
      }
    }),
    [address, chains, client]
  );
  return { adapter };
}
export {
  useEvmWagmiAdapter
};
