// @dialectlabs/blinks/dist/hooks/evm/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/hooks/evm/useEvmWagmiAdapter.ts
require('@dialectlabs/blinks-core');
var _react = require('react');
var _wagmi = require('wagmi');


var _viem = require('viem');




var _actions = require('viem/actions');
function useEvmWagmiAdapter({
  onConnectWalletRequest
}) {
  const chains = _wagmi.useChains.call(void 0, );
  const { address } = _wagmi.useAccount.call(void 0, );
  const { data: client } = _wagmi.useConnectorClient.call(void 0, );
  const adapter = _react.useMemo.call(void 0, 
    () => ({
      metadata: {
        supportedBlockchainIds: chains.map((chain) => `eip155:${chain.id}`)
      },
      connect: async () => {
        if (!address) {
          try {
            const potentialAddress = await onConnectWalletRequest();
            return potentialAddress || null;
          } catch (error) {
            console.error("Blink wagmi adapter connection failed:", error);
            return null;
          }
        }
        return address;
      },
      signTransaction: async (tx) => {
        if (!address || !client) {
          return { error: "No wallet connected" };
        }
        try {
          const deserializedTx = _wagmi.deserialize.call(void 0, tx);
          const hash = await _actions.sendTransaction.call(void 0, client, {
            ...deserializedTx,
            account: address
          });
          return { signature: hash };
        } catch (e) {
          console.error("Blink wagmi adapter sign transaction failed:", e);
          if (e instanceof _viem.TransactionExecutionError) {
            return { error: `Unable to sign: ${e.shortMessage}` };
          }
          return { error: "Signing failed" };
        }
      },
      signMessage: async (data) => {
        if (!address || !client) {
          return { error: "No wallet connected" };
        }
        let message;
        if (typeof data === "string") {
          message = data;
        } else if (typeof data === "object" && "raw" in data) {
          message = data;
        } else {
          message = JSON.stringify(data);
        }
        try {
          const signature = await _actions.signMessage.call(void 0, client, {
            account: address,
            message
          });
          return { signature };
        } catch (e) {
          console.error("Blink wagmi adapter message signing failed:", e);
          return { error: "Message signing failed" };
        }
      },
      confirmTransaction: async (signature) => {
        if (!client) {
          throw new Error("No wallet connected");
        }
        const res = await _actions.waitForTransactionReceipt.call(void 0, client, {
          hash: signature
        });
        if (res.status !== "success") {
          throw new Error("Transaction confirmation failed");
        }
      }
    }),
    [address, chains, client]
  );
  return { adapter };
}


exports.useEvmWagmiAdapter = useEvmWagmiAdapter;
