// @dialectlabs/blinks-core/dist/chunk-ROVRYL42.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }










var _chunkXNEBNDEXcjs = require('./chunk-XNEBNDEX.cjs');

// src/service-api/useBlink.ts
var _react = require('react');

// src/hooks/useBlink.ts


// src/hooks/useBlinkApiUrl.ts

function useBlinkApiUrl(url) {
  const [apiUrl, setApiUrl] = _react.useState.call(void 0, null);
  const [isLoading, setIsLoading] = _react.useState.call(void 0, false);
  _react.useEffect.call(void 0, () => {
    let ignore = false;
    if (!url) {
      setApiUrl(null);
      return;
    }
    setIsLoading(true);
    _chunkXNEBNDEXcjs.unfurlUrlToBlinkApiUrl.call(void 0, new URL(url)).then((apiUrl2) => {
      if (ignore) {
        return;
      }
      setIsLoading(false);
      setApiUrl(apiUrl2);
    }).catch((e) => {
      console.error(
        "[@dialectlabs/blinks-core] Failed to unfurl blink URL",
        e
      );
      if (!ignore) {
        setApiUrl(null);
      }
    }).finally(() => {
      if (!ignore) {
        setIsLoading(false);
      }
    });
    return () => {
      ignore = true;
    };
  }, [url]);
  return { actionApiUrl: apiUrl, blinkApiUrl: apiUrl, isUrlLoading: isLoading };
}

// src/hooks/useBlinksRegistryInterval.ts

function useBlinksRegistryInterval() {
  const [isRegistryLoaded, setRegistryLoaded] = _react.useState.call(void 0, false);
  _react.useEffect.call(void 0, () => {
    const registry = _chunkXNEBNDEXcjs.BlinksRegistry.getInstance();
    registry.init().then(() => {
      setRegistryLoaded(true);
    });
    return () => {
      registry.stopRefresh();
    };
  }, [isRegistryLoaded]);
  return { isRegistryLoaded };
}

// src/hooks/useBlink.ts
function useBlink2({
  url,
  supportStrategy = _chunkXNEBNDEXcjs.defaultBlinkSupportStrategy
}) {
  const { isRegistryLoaded } = useBlinksRegistryInterval();
  const {
    blink: blinkJson,
    isLoading,
    metadata,
    refresh,
    blinkApiUrl,
    originalBlinkApiUrl
  } = useBlink({ url: url.toString() });
  const blink = _react.useMemo.call(void 0, 
    () => isRegistryLoaded && blinkApiUrl && blinkJson && metadata ? _chunkXNEBNDEXcjs.BlinkInstance.hydrate(
      blinkApiUrl,
      blinkJson,
      metadata,
      supportStrategy
    ) : null,
    [blinkApiUrl, blinkJson, metadata, supportStrategy, isRegistryLoaded]
  );
  return {
    blink,
    blinkJson,
    metadata,
    blinkApiUrl,
    originalBlinkApiUrl,
    isLoading: isLoading || !isRegistryLoaded,
    refresh
  };
}
function useAction({
  url,
  supportStrategy = _chunkXNEBNDEXcjs.defaultBlinkSupportStrategy
}) {
  const { isRegistryLoaded } = useBlinksRegistryInterval();
  const { blinkApiUrl, isUrlLoading } = useBlinkApiUrl(url);
  const [blink, setBlink] = _react.useState.call(void 0, null);
  const [isLoading, setIsLoading] = _react.useState.call(void 0, false);
  const [hasFetched, setHasFetched] = _react.useState.call(void 0, false);
  const fetchAction = _react.useCallback.call(void 0, () => {
    if (!blinkApiUrl) {
      return () => {
      };
    }
    let ignore = false;
    setIsLoading(true);
    setHasFetched(false);
    _chunkXNEBNDEXcjs.BlinkInstance.fetch(blinkApiUrl, supportStrategy).then((action) => {
      if (!ignore) {
        setIsLoading(false);
        setBlink(action);
        setHasFetched(true);
      }
    }).catch((e) => {
      if (!ignore) {
        console.error("[@dialectlabs/blinks-core] Failed to fetch action", e);
        setBlink(null);
      }
    }).finally(() => {
      if (!ignore) {
        setIsLoading(false);
      }
    });
    return () => {
      ignore = true;
    };
  }, [blinkApiUrl, supportStrategy]);
  _react.useEffect.call(void 0, () => {
    if (!isRegistryLoaded) {
      return;
    }
    const cleanup = fetchAction();
    return () => {
      cleanup();
    };
  }, [blinkApiUrl, isRegistryLoaded]);
  _react.useEffect.call(void 0, () => {
    if (!blink || !hasFetched) {
      return;
    }
    try {
      const updated = blink.withUpdate({
        supportStrategy
      });
      setBlink(updated);
    } catch (e) {
      console.error("[@dialectlabs/blinks-core] Failed to update action", e);
    }
  }, [supportStrategy, hasFetched]);
  return {
    // NOTE: this will be removed in the nearest future
    action: blink,
    blink,
    isLoading: !isRegistryLoaded || isUrlLoading || isLoading,
    refresh: fetchAction
  };
}

// src/hooks/useBlinkList.ts

var useBlinkList = ({ id, wallet }) => {
  const [loading, setLoading] = _react.useState.call(void 0, false);
  const [data, setData] = _react.useState.call(void 0, );
  const refetch = _react.useCallback.call(void 0, () => {
    let ignore = false;
    setLoading(true);
    fetchBlinkList(id, wallet).then((data2) => {
      if (!ignore) {
        setData(data2);
      }
    }).finally(() => {
      if (!ignore) {
        setLoading(false);
      }
    });
    return () => {
      ignore = true;
    };
  }, [id, wallet]);
  _react.useEffect.call(void 0, () => {
    const cancel = refetch();
    return () => {
      cancel();
    };
  }, [refetch]);
  return {
    loading,
    refetch,
    data: _nullishCoalesce(_optionalChain([data, 'optionalAccess', _ => _.entries]), () => ( []))
  };
};
async function fetchBlinkList(id, wallet) {
  try {
    const urlObj = new URL(
      `https://registry.dial.to/v1/private/blink-lists/${id}`
    );
    if (wallet) {
      urlObj.searchParams.append("account", wallet);
    }
    const response = await fetch(urlObj, {
      method: "GET",
      headers: {
        Accept: "application/json",
        ..._chunkXNEBNDEXcjs.clientKey && { [_chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER]: _chunkXNEBNDEXcjs.clientKey }
      }
    });
    if (!response.ok) {
      console.error(
        `[@dialectlabs/blinks] Failed to fetch blink list, response status: ${response.status}`
      );
      return {
        entries: []
      };
    }
    return await response.json();
  } catch (e) {
    console.error(`[@dialectlabs/blinks] Failed to fetch blink list`, e);
    return {
      entries: []
    };
  }
}

// src/hooks/useMetadata.ts

var useMetadata = ({ url, wallet }) => {
  const [loading, setLoading] = _react.useState.call(void 0, false);
  const [data, setData] = _react.useState.call(void 0, );
  const refetch = _react.useCallback.call(void 0, () => {
    let ignore = false;
    setLoading(true);
    fetchMetadata(url, wallet).then((data2) => {
      if (!ignore) {
        setData(data2);
      }
    }).finally(() => {
      if (!ignore) {
        setLoading(false);
      }
    });
    return () => {
      ignore = true;
    };
  }, [url, wallet]);
  _react.useEffect.call(void 0, () => {
    const cancel = refetch();
    return () => {
      cancel();
    };
  }, [refetch]);
  return {
    loading,
    refetch,
    data
  };
};
async function fetchMetadata(url, wallet) {
  try {
    const urlObj = new URL(url);
    if (wallet) {
      urlObj.searchParams.append("account", wallet);
    }
    const { url: proxyUrl, headers: proxyHeaders } = _chunkXNEBNDEXcjs.proxifyMetadata.call(void 0, 
      urlObj.toString()
    );
    const response = await fetch(proxyUrl, {
      method: "GET",
      headers: {
        Accept: "application/json",
        ...proxyHeaders
      }
    });
    if (!response.ok) {
      console.error(
        `[@dialectlabs/blinks] Failed to fetch metadata, response status: ${response.status}`
      );
      return {
        rows: []
      };
    }
    return await response.json();
  } catch (e) {
    console.error(`[@dialectlabs/blinks] Failed to fetch metadata`, e);
    return {
      rows: []
    };
  }
}

// src/service-api/useBlink.ts
var useBlink = ({ url }) => {
  const [blinkData, setBlinkData] = _react.useState.call(void 0, null);
  const [isLoading, setIsLoading] = _react.useState.call(void 0, false);
  const { blinkApiUrl, isUrlLoading } = useBlinkApiUrl(url);
  const loadBlink = _react.useCallback.call(void 0, 
    async ({ abortController } = {}) => {
      if (!url || !blinkApiUrl) {
        return;
      }
      setIsLoading(true);
      try {
        const blinkData2 = await fetchBlink(blinkApiUrl);
        if (!_optionalChain([abortController, 'optionalAccess', _2 => _2.signal, 'access', _3 => _3.aborted])) {
          setBlinkData(blinkData2);
          setIsLoading(false);
        }
      } catch (e2) {
        if (!_optionalChain([abortController, 'optionalAccess', _4 => _4.signal, 'access', _5 => _5.aborted])) {
          console.error(
            "[@dialectlabs/blinks-core] Failed to fetch blink",
            blinkApiUrl
          );
          setBlinkData(null);
          setIsLoading(false);
        }
      }
    },
    [url, blinkApiUrl]
  );
  _react.useEffect.call(void 0, () => {
    const abortController = new AbortController();
    loadBlink({ abortController });
    return () => {
      abortController.abort();
    };
  }, [loadBlink]);
  return {
    // raw blink data
    blink: _nullishCoalesce(_optionalChain([blinkData, 'optionalAccess', _6 => _6.blink]), () => ( null)),
    blinkApiUrl: _nullishCoalesce(_optionalChain([blinkData, 'optionalAccess', _7 => _7.url]), () => ( null)),
    originalBlinkApiUrl: blinkApiUrl,
    // blink metadata, such as blockchain and solana actions version (if specified in blink response)
    metadata: _nullishCoalesce(_optionalChain([blinkData, 'optionalAccess', _8 => _8.metadata]), () => ( null)),
    isLoading: isLoading || isUrlLoading,
    refresh: loadBlink
  };
};
var fetchBlink = async (apiUrl, options = {}) => {
  let url;
  if (_chunkXNEBNDEXcjs.shouldIgnoreProxy.call(void 0, new URL(apiUrl)) || _chunkXNEBNDEXcjs.shouldPreserveOriginal.call(void 0, new URL(apiUrl))) {
    url = new URL(apiUrl);
  } else {
    url = new URL(`https://api.dial.to/v1/blink`);
    url.searchParams.append("apiUrl", apiUrl);
  }
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      ..._chunkXNEBNDEXcjs.clientKey && { [_chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER]: _chunkXNEBNDEXcjs.clientKey }
    },
    signal: _optionalChain([options, 'access', _9 => _9.abortController, 'optionalAccess', _10 => _10.signal])
  });
  if (!response.ok) {
    throw new Error("Failed to fetch blink");
  }
  const blink = await response.json();
  const metadata = _chunkXNEBNDEXcjs.getBlinkSupportabilityMetadata.call(void 0, response);
  return { blink, metadata, url: url.toString() };
};

// src/service-api/useBlinkDataTable.ts

var useBlinkDataTable = ({
  blinkApiUrl,
  wallet
}) => {
  const [isLoading, setLoading] = _react.useState.call(void 0, false);
  const [data, setData] = _react.useState.call(void 0, null);
  const loadBlinkDataTable = _react.useCallback.call(void 0, 
    async (options = {}) => {
      if (!blinkApiUrl) {
        return;
      }
      setLoading(true);
      try {
        const data2 = await fetchBlinkDataTable(blinkApiUrl, wallet, options);
        if (!_optionalChain([options, 'access', _11 => _11.abortController, 'optionalAccess', _12 => _12.signal, 'access', _13 => _13.aborted])) {
          setData(data2);
          setLoading(false);
        }
      } catch (e3) {
        if (!_optionalChain([options, 'access', _14 => _14.abortController, 'optionalAccess', _15 => _15.signal, 'access', _16 => _16.aborted])) {
          console.error(
            "[@dialectlabs/blinks-core] Failed to fetch blink data table",
            blinkApiUrl
          );
          setData(null);
          setLoading(false);
        }
      }
    },
    [blinkApiUrl, wallet]
  );
  _react.useEffect.call(void 0, () => {
    const abortController = new AbortController();
    loadBlinkDataTable({ abortController });
    return () => {
      abortController.abort();
    };
  }, [loadBlinkDataTable]);
  return {
    data,
    isLoading,
    refresh: loadBlinkDataTable
  };
};
async function fetchBlinkDataTable(blinkApiUrl, wallet, options = {}) {
  const url = new URL("https://api.dial.to/v1/blink-data-table");
  url.searchParams.append("apiUrl", blinkApiUrl);
  if (wallet) {
    url.searchParams.append("account", wallet);
  }
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      ..._chunkXNEBNDEXcjs.clientKey && { [_chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER]: _chunkXNEBNDEXcjs.clientKey }
    },
    signal: _optionalChain([options, 'access', _17 => _17.abortController, 'optionalAccess', _18 => _18.signal])
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch blink data table, url: ${blinkApiUrl}`);
  }
  return response.json();
}

// src/service-api/useBlinkList.ts

var useBlinkList2 = ({
  id,
  wallet
}) => {
  const [loading, setLoading] = _react.useState.call(void 0, false);
  const [data, setData] = _react.useState.call(void 0, null);
  const loadBlinkList = _react.useCallback.call(void 0, 
    async (options = {}) => {
      if (!id) {
        return;
      }
      setLoading(true);
      try {
        const data2 = await fetchBlinkList2(id, wallet, options);
        if (!_optionalChain([options, 'access', _19 => _19.abortController, 'optionalAccess', _20 => _20.signal, 'access', _21 => _21.aborted])) {
          setData(data2);
          setLoading(false);
        }
      } catch (e4) {
        if (!_optionalChain([options, 'access', _22 => _22.abortController, 'optionalAccess', _23 => _23.signal, 'access', _24 => _24.aborted])) {
          console.error(
            `[@dialectlabs/blinks-core] Failed to fetch blink list: ${id}`
          );
          setData(null);
          setLoading(false);
        }
      }
    },
    [id, wallet]
  );
  _react.useEffect.call(void 0, () => {
    const abortController = new AbortController();
    loadBlinkList({ abortController });
    return () => {
      abortController.abort();
    };
  }, [loadBlinkList]);
  return {
    loading,
    refresh: loadBlinkList,
    data
  };
};
async function fetchBlinkList2(id, wallet, options = {}) {
  const urlObj = new URL(`https://api.dial.to/v1/blink-lists/${id}`);
  if (wallet) {
    urlObj.searchParams.append("account", wallet);
  }
  const response = await fetch(urlObj, {
    method: "GET",
    headers: {
      Accept: "application/json",
      ..._chunkXNEBNDEXcjs.clientKey && { [_chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER]: _chunkXNEBNDEXcjs.clientKey }
    },
    signal: _optionalChain([options, 'access', _25 => _25.abortController, 'optionalAccess', _26 => _26.signal])
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch blink list: ${id}`);
  }
  return response.json();
}

// src/service-api/useBlinkPreview.ts

var useBlinkPreview = ({
  blinkApiUrl
}) => {
  const [isLoading, setLoading] = _react.useState.call(void 0, false);
  const [preview, setPreview] = _react.useState.call(void 0, null);
  const loadBlinkPreview = _react.useCallback.call(void 0, 
    async (options = {}) => {
      if (!blinkApiUrl) {
        return;
      }
      setLoading(true);
      try {
        const data = await fetchBlinkPreview(blinkApiUrl, options);
        if (!_optionalChain([options, 'access', _27 => _27.abortController, 'optionalAccess', _28 => _28.signal, 'access', _29 => _29.aborted])) {
          setPreview(data);
          setLoading(false);
        }
      } catch (e5) {
        if (!_optionalChain([options, 'access', _30 => _30.abortController, 'optionalAccess', _31 => _31.signal, 'access', _32 => _32.aborted])) {
          console.error(
            "[@dialectlabs/blinks-core] Failed to fetch blink preview",
            blinkApiUrl
          );
          setPreview(null);
          setLoading(false);
        }
      }
    },
    [blinkApiUrl]
  );
  _react.useEffect.call(void 0, () => {
    const abortController = new AbortController();
    loadBlinkPreview({ abortController });
    return () => {
      abortController.abort();
    };
  }, [loadBlinkPreview]);
  return {
    preview,
    isLoading,
    refresh: loadBlinkPreview
  };
};
async function fetchBlinkPreview(blinkApiUrl, options = {}) {
  const url = new URL("https://api.dial.to/v1/blink-preview");
  url.searchParams.append("apiUrl", blinkApiUrl);
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      ..._chunkXNEBNDEXcjs.clientKey && { [_chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER]: _chunkXNEBNDEXcjs.clientKey }
    },
    signal: _optionalChain([options, 'access', _33 => _33.abortController, 'optionalAccess', _34 => _34.signal])
  });
  if (!response.ok) {
    throw new Error(`Failed to fetch blink preview, url: ${blinkApiUrl}`);
  }
  return response.json();
}


















exports.useBlink = useBlink; exports.fetchBlink = fetchBlink; exports.useBlinkDataTable = useBlinkDataTable; exports.fetchBlinkDataTable = fetchBlinkDataTable; exports.useBlinkList = useBlinkList2; exports.fetchBlinkList = fetchBlinkList2; exports.useBlinkPreview = useBlinkPreview; exports.fetchBlinkPreview = fetchBlinkPreview; exports.useBlinkApiUrl = useBlinkApiUrl; exports.useBlinksRegistryInterval = useBlinksRegistryInterval; exports.useBlink2 = useBlink2; exports.useAction = useAction; exports.useBlinkList2 = useBlinkList; exports.fetchBlinkList2 = fetchBlinkList; exports.useMetadata = useMetadata; exports.fetchMetadata = fetchMetadata;
