// @dialectlabs/blinks-core/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }








var _chunkROVRYL42cjs = require('./chunk-ROVRYL42.cjs');








































var _chunkXNEBNDEXcjs = require('./chunk-XNEBNDEX.cjs');

// src/BlinkContainer.tsx





var _react = require('react');

// src/utils/type-guards.ts
var isSignTransactionError = (data) => !!data.error;
var isSignMessageError = isSignTransactionError;
var isPostRequestError = (data) => !!data.error;

// src/utils/validators.ts
var isURL = (url) => {
  try {
    new URL(url);
    return true;
  } catch (e2) {
    return false;
  }
};

// src/BlinkContainer.tsx
var _jsxruntime = require('react/jsx-runtime');
var DisclaimerType = /* @__PURE__ */ ((DisclaimerType2) => {
  DisclaimerType2["BLOCKED"] = "blocked";
  DisclaimerType2["UNKNOWN"] = "unknown";
  return DisclaimerType2;
})(DisclaimerType || {});
var ExecutionType = /* @__PURE__ */ ((ExecutionType2) => {
  ExecutionType2["CHECK_SUPPORTABILITY"] = "CHECK_SUPPORTABILITY";
  ExecutionType2["INITIATE"] = "INITIATE";
  ExecutionType2["FINISH"] = "FINISH";
  ExecutionType2["FAIL"] = "FAIL";
  ExecutionType2["RESET"] = "RESET";
  ExecutionType2["SOFT_RESET"] = "SOFT_RESET";
  ExecutionType2["UNBLOCK"] = "UNBLOCK";
  ExecutionType2["BLOCK"] = "BLOCK";
  ExecutionType2["PARTIAL_UPDATE"] = "PARTIAL_UPDATE";
  return ExecutionType2;
})(ExecutionType || {});
var executionReducer = (state, action) => {
  switch (action.type) {
    case "CHECK_SUPPORTABILITY" /* CHECK_SUPPORTABILITY */:
      return {
        status: "checking-supportability",
        checkingSupportability: true
      };
    case "INITIATE" /* INITIATE */:
      return { status: "executing", executingAction: action.executingAction };
    case "FINISH" /* FINISH */:
      return {
        ...state,
        status: "success",
        successMessage: action.successMessage,
        errorMessage: null
      };
    case "FAIL" /* FAIL */:
      return {
        ...state,
        status: "error",
        errorMessage: action.errorMessage,
        successMessage: null
      };
    case "RESET" /* RESET */:
      return {
        status: "idle"
      };
    case "SOFT_RESET" /* SOFT_RESET */:
      return {
        ...state,
        executingAction: null,
        status: "idle",
        errorMessage: action.errorMessage,
        successMessage: null
      };
    case "BLOCK" /* BLOCK */:
      return {
        status: "blocked"
      };
    case "UNBLOCK" /* UNBLOCK */:
      return {
        status: "idle"
      };
    case "PARTIAL_UPDATE" /* PARTIAL_UPDATE */:
      return {
        ...state,
        generalMessage: action.message
      };
  }
};
var getOverallBlinkState = (blink, websiteUrl) => {
  const blinkState = _chunkXNEBNDEXcjs.getExtendedBlinkState.call(void 0, blink);
  const originalUrlData = websiteUrl ? _chunkXNEBNDEXcjs.isInterstitial.call(void 0, websiteUrl) : null;
  if (!originalUrlData) {
    return {
      blink: blinkState
    };
  }
  if (originalUrlData.isInterstitial) {
    return {
      blink: blinkState,
      origin: _chunkXNEBNDEXcjs.getExtendedInterstitialState.call(void 0, websiteUrl),
      originType: "interstitials"
    };
  }
  return {
    blink: blinkState,
    origin: _chunkXNEBNDEXcjs.getExtendedWebsiteState.call(void 0, websiteUrl),
    originType: "websites"
  };
};
var checkSecurityFromBlinkState = (state, normalizedSecurityLevel) => {
  const checkBlink = _chunkXNEBNDEXcjs.checkSecurity.call(void 0, 
    state.blink,
    normalizedSecurityLevel.actions
  );
  if (!state.origin) {
    return checkBlink;
  }
  return checkBlink && _chunkXNEBNDEXcjs.checkSecurity.call(void 0, state.origin, normalizedSecurityLevel[state.originType]);
};
var DEFAULT_SECURITY_LEVEL = "only-trusted";
var BlinkContainer = ({
  blink: initialBlink,
  adapter,
  websiteUrl,
  websiteText,
  callbacks = _chunkXNEBNDEXcjs.EMPTY_OBJECT,
  securityLevel = DEFAULT_SECURITY_LEVEL,
  Layout,
  selector
}) => {
  const [blink, setBlink] = _react.useState.call(void 0, initialBlink);
  const singleComponent = _react.useMemo.call(void 0, () => _optionalChain([selector, 'optionalCall', _ => _(blink)]), [blink, selector]);
  const isPartialBlink = typeof selector === "function";
  const normalizedSecurityLevel = _react.useMemo.call(void 0, () => {
    if (typeof securityLevel === "string") {
      return {
        websites: securityLevel,
        interstitials: securityLevel,
        actions: securityLevel
      };
    }
    return securityLevel;
  }, [securityLevel]);
  const [blinkState, setBlinkState] = _react.useState.call(void 0, 
    getOverallBlinkState(blink, websiteUrl)
  );
  const [supportability, setSupportability] = _react.useState.call(void 0, {
    isSupported: true
  });
  const overallState = _react.useMemo.call(void 0, 
    () => _chunkXNEBNDEXcjs.mergeBlinkStates.call(void 0, 
      ...[blinkState.blink, blinkState.origin].filter(
        Boolean
      )
    ),
    [blinkState]
  );
  const isPassingSecurityCheck = isPartialBlink ? true : checkSecurityFromBlinkState(blinkState, normalizedSecurityLevel);
  const [executionState, dispatch] = _react.useReducer.call(void 0, executionReducer, {
    status: isPartialBlink ? "idle" : "checking-supportability"
  });
  _react.useEffect.call(void 0, () => {
    if (blink === initialBlink) {
      return;
    }
    setBlink(initialBlink);
    setBlinkState(getOverallBlinkState(initialBlink, websiteUrl));
    dispatch({
      type: isPartialBlink ? "RESET" /* RESET */ : "CHECK_SUPPORTABILITY" /* CHECK_SUPPORTABILITY */
    });
  }, [initialBlink, websiteUrl]);
  _react.useEffect.call(void 0, () => {
    _optionalChain([callbacks, 'access', _2 => _2.onActionMount, 'optionalCall', _3 => _3(
      initialBlink,
      _nullishCoalesce(websiteUrl, () => ( initialBlink.url)),
      blinkState.blink
    )]);
  }, []);
  _react.useEffect.call(void 0, () => {
    const liveDataConfig = blink.liveData_experimental;
    if (!liveDataConfig || !liveDataConfig.enabled || executionState.status !== "idle" || blink.isChained) {
      return;
    }
    let timeout;
    const fetcher = async () => {
      try {
        const newBlink = await blink.refresh();
        if (executionState.status === "idle") {
          setBlink(newBlink);
        }
      } catch (e) {
        console.error(
          `[@dialectlabs/blinks] Failed to fetch live data for blink ${blink.url}`
        );
        timeout = setTimeout(fetcher, liveDataConfig.delayMs);
      }
    };
    timeout = setTimeout(fetcher, liveDataConfig.delayMs);
    return () => {
      clearTimeout(timeout);
    };
  }, [blink, executionState.status, isPartialBlink]);
  _react.useEffect.call(void 0, () => {
    const checkSupportability = async (blink2) => {
      if (blink2.isChained || executionState.status !== "checking-supportability") {
        return;
      }
      try {
        const supportability2 = await blink2.isSupported(adapter);
        setSupportability(supportability2);
      } finally {
        dispatch({
          type: overallState !== "malicious" && isPassingSecurityCheck ? "RESET" /* RESET */ : "BLOCK" /* BLOCK */
        });
      }
    };
    checkSupportability(blink);
  }, [
    blink,
    adapter,
    executionState.status,
    overallState,
    isPassingSecurityCheck
  ]);
  const execute = async (component, params) => {
    if (params) {
      if (component instanceof _chunkXNEBNDEXcjs.FormActionComponent) {
        Object.entries(params).forEach(
          ([name, value]) => component.setValue(value, name)
        );
      }
      if (component instanceof _chunkXNEBNDEXcjs.MultiValueActionComponent) {
        component.setValue(params[component.parameter.name]);
      }
      if (component instanceof _chunkXNEBNDEXcjs.SingleValueActionComponent) {
        const incomingValues = params[component.parameter.name];
        const value = typeof incomingValues === "string" ? incomingValues : incomingValues[0];
        component.setValue(value);
      }
    }
    const newBlinkState = getOverallBlinkState(blink, websiteUrl);
    const newIsPassingSecurityCheck = checkSecurityFromBlinkState(
      newBlinkState,
      normalizedSecurityLevel
    );
    if ((newBlinkState.blink !== blinkState.blink || newBlinkState.origin !== blinkState.origin) && !newIsPassingSecurityCheck) {
      setBlinkState(newBlinkState);
      dispatch({ type: "BLOCK" /* BLOCK */ });
      _optionalChain([callbacks, 'access', _4 => _4.onActionCancel, 'optionalCall', _5 => _5(blink, component, "security-state-changed")]);
      return;
    }
    if (component.type === "inline-link") {
      if (isURL(component.href)) {
        return {
          type: "external-link",
          data: {
            externalLink: component.href
          },
          onNext: () => {
          },
          onCancel: () => {
          }
        };
      }
      return;
    }
    dispatch({ type: "INITIATE" /* INITIATE */, executingAction: component });
    const context = {
      action: component.parent,
      actionType: blinkState.blink,
      originalUrl: _nullishCoalesce(websiteUrl, () => ( component.parent.url)),
      triggeredLinkedAction: component
    };
    try {
      const account = await adapter.connect(context);
      if (!account) {
        dispatch({ type: "RESET" /* RESET */ });
        _optionalChain([callbacks, 'optionalAccess', _6 => _6.onActionCancel, 'optionalCall', _7 => _7(blink, component, "wallet-not-connected")]);
        return;
      }
      const response = await component.post(account).catch((e) => ({ error: e.message }));
      if (isPostRequestError(response)) {
        dispatch({
          type: "SOFT_RESET" /* SOFT_RESET */,
          errorMessage: isPostRequestError(response) ? response.error : "Transaction data missing"
        });
        _optionalChain([callbacks, 'access', _8 => _8.onActionError, 'optionalCall', _9 => _9(blink, component, "post-request-error")]);
        return;
      }
      if (_optionalChain([response, 'access', _10 => _10.lifecycle, 'optionalAccess', _11 => _11.executing])) {
        dispatch({
          type: "PARTIAL_UPDATE" /* PARTIAL_UPDATE */,
          message: response.lifecycle.executing.message
        });
      }
      const chain = async (signature) => {
        if (!_optionalChain([response, 'access', _12 => _12.links, 'optionalAccess', _13 => _13.next]) && !_optionalChain([response, 'access', _14 => _14.lifecycle, 'optionalAccess', _15 => _15.success])) {
          dispatch({
            type: "FINISH" /* FINISH */,
            successMessage: response.message
          });
          _optionalChain([callbacks, 'access', _16 => _16.onActionComplete, 'optionalCall', _17 => _17(blink, component, signature)]);
          return;
        }
        if (response.type === "message" && !signature) {
          dispatch({
            type: "SOFT_RESET" /* SOFT_RESET */,
            errorMessage: _nullishCoalesce(_optionalChain([response, 'access', _18 => _18.lifecycle, 'optionalAccess', _19 => _19.error, 'optionalAccess', _20 => _20.message]), () => ( "Missing signature for message"))
          });
          _optionalChain([callbacks, 'access', _21 => _21.onActionError, 'optionalCall', _22 => _22(
            blink,
            component,
            "message-signature-missing"
          )]);
          return;
        }
        const chainData = response.type === "message" ? {
          signature,
          account,
          state: response.state,
          data: response.data
        } : {
          signature,
          account
        };
        const nextBlink = _optionalChain([response, 'access', _23 => _23.links, 'optionalAccess', _24 => _24.next]) ? await blink.chain(
          response.links.next,
          chainData,
          _optionalChain([response, 'access', _25 => _25.lifecycle, 'optionalAccess', _26 => _26.success])
        ) : blink.safeInlineChain(_optionalChain([response, 'access', _27 => _27.lifecycle, 'optionalAccess', _28 => _28.success]));
        if (!nextBlink || isPartialBlink && !_optionalChain([selector, 'optionalCall', _29 => _29(nextBlink)])) {
          dispatch({
            type: "FINISH" /* FINISH */,
            successMessage: response.message
          });
          _optionalChain([callbacks, 'access', _30 => _30.onActionComplete, 'optionalCall', _31 => _31(blink, component, signature)]);
          return;
        }
        setBlink(nextBlink);
        dispatch({ type: "RESET" /* RESET */ });
        _optionalChain([callbacks, 'access', _32 => _32.onActionChain, 'optionalCall', _33 => _33(
          blink,
          nextBlink,
          component,
          response.type,
          signature
        )]);
      };
      if (response.type === "transaction" || !response.type) {
        const signResult = await adapter.signTransaction(
          response.transaction,
          context
        );
        if (!signResult || isSignTransactionError(signResult)) {
          dispatch({
            type: "SOFT_RESET" /* SOFT_RESET */,
            errorMessage: _nullishCoalesce(signResult.error, () => ( "Unable to sign transaction"))
          });
          _optionalChain([callbacks, 'access', _34 => _34.onActionCancel, 'optionalCall', _35 => _35(
            blink,
            component,
            "transaction-sign-cancel"
          )]);
          return;
        }
        const confirmationResult = await adapter.confirmTransaction(signResult.signature, context).then(() => ({ success: true })).catch((e) => ({ success: false, message: e.message }));
        if (!confirmationResult.success) {
          dispatch({
            type: "SOFT_RESET" /* SOFT_RESET */,
            errorMessage: _nullishCoalesce(_nullishCoalesce(_optionalChain([response, 'access', _36 => _36.lifecycle, 'optionalAccess', _37 => _37.error, 'optionalAccess', _38 => _38.message]), () => ( confirmationResult.message)), () => ( "Unknown error, please try again"))
          });
          _optionalChain([callbacks, 'access', _39 => _39.onActionError, 'optionalCall', _40 => _40(
            blink,
            component,
            "transaction-confirmation-failed",
            signResult.signature
          )]);
          return;
        }
        await chain(signResult.signature);
        return;
      }
      if (response.type === "message") {
        const signResult = await adapter.signMessage(response.data, context);
        if (!signResult || isSignMessageError(signResult)) {
          dispatch({ type: "RESET" /* RESET */ });
          _optionalChain([callbacks, 'access', _41 => _41.onActionCancel, 'optionalCall', _42 => _42(blink, component, "message-sign-cancel")]);
          return;
        }
        await chain(signResult.signature);
        return;
      }
      if (response.type === "post") {
        await chain();
        return;
      }
      if (response.type === "external-link") {
        if (isURL(response.externalLink)) {
          return {
            type: "external-link",
            data: {
              externalLink: response.externalLink
            },
            onNext: () => chain(),
            onCancel: () => chain()
          };
        }
        await chain();
        return;
      }
    } catch (e) {
      dispatch({
        type: "SOFT_RESET" /* SOFT_RESET */,
        errorMessage: _nullishCoalesce(e.message, () => ( "Unknown error, please try again"))
      });
      _optionalChain([callbacks, 'access', _43 => _43.onActionError, 'optionalCall', _44 => _44(blink, component, "unknown-error")]);
    }
  };
  const disclaimer = _react.useMemo.call(void 0, () => {
    if (overallState === "malicious") {
      return {
        type: "blocked" /* BLOCKED */,
        ignorable: isPassingSecurityCheck,
        hidden: executionState.status !== "blocked" && executionState.status !== "checking-supportability",
        onSkip: () => dispatch({ type: "UNBLOCK" /* UNBLOCK */ })
      };
    }
    if (overallState === "unknown") {
      return {
        type: "unknown" /* UNKNOWN */,
        ignorable: isPassingSecurityCheck
      };
    }
    return null;
  }, [executionState.status, isPassingSecurityCheck, overallState]);
  const blinkCaption = _react.useMemo.call(void 0, () => {
    const error = _nullishCoalesce(executionState.errorMessage, () => ( blink.error));
    const generalMessage = _nullishCoalesce(executionState.generalMessage, () => ( blink.message));
    if (error) {
      return { type: "error", text: error };
    }
    if (executionState.successMessage) {
      return { type: "success", text: executionState.successMessage };
    }
    if (generalMessage) {
      return { type: "default", text: generalMessage };
    }
    return null;
  }, [
    executionState.errorMessage,
    executionState.successMessage,
    executionState.generalMessage,
    blink.error,
    blink.message
  ]);
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
    Layout,
    {
      securityState: overallState,
      websiteUrl,
      websiteText,
      blink,
      component: singleComponent,
      caption: blinkCaption,
      executionStatus: executionState.status,
      executingAction: executionState.executingAction,
      executeFn: execute,
      disclaimer,
      supportability,
      id: blink.id
    }
  );
};




























































exports.AbstractActionComponent = _chunkXNEBNDEXcjs.AbstractActionComponent; exports.Action = _chunkXNEBNDEXcjs.BlinkInstance; exports.ActionsRegistry = _chunkXNEBNDEXcjs.BlinksRegistry; exports.ActionsURLMapper = _chunkXNEBNDEXcjs.BlinksURLMapper; exports.BASELINE_ACTION_BLOCKCHAIN_IDS = _chunkXNEBNDEXcjs.BASELINE_BLINK_BLOCKCHAIN_IDS; exports.BASELINE_ACTION_VERSION = _chunkXNEBNDEXcjs.BASELINE_ACTION_VERSION; exports.BASELINE_BLINK_BLOCKCHAIN_IDS = _chunkXNEBNDEXcjs.BASELINE_BLINK_BLOCKCHAIN_IDS; exports.BLINK_CLIENT_KEY_HEADER = _chunkXNEBNDEXcjs.BLINK_CLIENT_KEY_HEADER; exports.BlinkContainer = BlinkContainer; exports.BlinkInstance = _chunkXNEBNDEXcjs.BlinkInstance; exports.BlinksRegistry = _chunkXNEBNDEXcjs.BlinksRegistry; exports.BlinksURLMapper = _chunkXNEBNDEXcjs.BlinksURLMapper; exports.BlockchainIds = _chunkXNEBNDEXcjs.BlockchainIds; exports.ButtonActionComponent = _chunkXNEBNDEXcjs.ButtonActionComponent; exports.DEFAULT_SUPPORTED_BLOCKCHAIN_IDS = _chunkXNEBNDEXcjs.DEFAULT_SUPPORTED_BLOCKCHAIN_IDS; exports.DisclaimerType = DisclaimerType; exports.ExecutionType = ExecutionType; exports.FormActionComponent = _chunkXNEBNDEXcjs.FormActionComponent; exports.MAX_SUPPORTED_ACTION_VERSION = _chunkXNEBNDEXcjs.MAX_SUPPORTED_ACTION_VERSION; exports.MultiValueActionComponent = _chunkXNEBNDEXcjs.MultiValueActionComponent; exports.SOLANA_ACTION_PREFIX = _chunkXNEBNDEXcjs.SOLANA_ACTION_PREFIX; exports.SignMessageVerificationErrorType = _chunkXNEBNDEXcjs.SignMessageVerificationErrorType; exports.SingleValueActionComponent = _chunkXNEBNDEXcjs.SingleValueActionComponent; exports.checkSecurity = _chunkXNEBNDEXcjs.checkSecurity; exports.createSignMessageText = _chunkXNEBNDEXcjs.createSignMessageText; exports.defaultActionSupportStrategy = _chunkXNEBNDEXcjs.defaultBlinkSupportStrategy; exports.defaultBlinkSupportStrategy = _chunkXNEBNDEXcjs.defaultBlinkSupportStrategy; exports.fetchActionsRegistryConfig = _chunkXNEBNDEXcjs.fetchBlinksRegistryConfig; exports.fetchBlinkList = _chunkROVRYL42cjs.fetchBlinkList2; exports.fetchMetadata = _chunkROVRYL42cjs.fetchMetadata; exports.getBlinkSupportabilityMetadata = _chunkXNEBNDEXcjs.getBlinkSupportabilityMetadata; exports.getExtendedActionState = _chunkXNEBNDEXcjs.getExtendedBlinkState; exports.getExtendedBlinkState = _chunkXNEBNDEXcjs.getExtendedBlinkState; exports.getExtendedInterstitialState = _chunkXNEBNDEXcjs.getExtendedInterstitialState; exports.getExtendedWebsiteState = _chunkXNEBNDEXcjs.getExtendedWebsiteState; exports.isBlockchainSupported = _chunkXNEBNDEXcjs.isBlockchainSupported; exports.isInterstitial = _chunkXNEBNDEXcjs.isInterstitial; exports.isParameterSelectable = _chunkXNEBNDEXcjs.isParameterSelectable; exports.isPatternAllowed = _chunkXNEBNDEXcjs.isPatternAllowed; exports.isProxified = _chunkXNEBNDEXcjs.isProxified; exports.isVersionSupported = _chunkXNEBNDEXcjs.isVersionSupported; exports.mergeActionStates = _chunkXNEBNDEXcjs.mergeBlinkStates; exports.mergeBlinkStates = _chunkXNEBNDEXcjs.mergeBlinkStates; exports.parseSignMessageText = _chunkXNEBNDEXcjs.parseSignMessageText; exports.proxify = _chunkXNEBNDEXcjs.proxify; exports.proxifyImage = _chunkXNEBNDEXcjs.proxifyImage; exports.setClientKey = _chunkXNEBNDEXcjs.setClientKey; exports.setProxyUrl = _chunkXNEBNDEXcjs.setProxyUrl; exports.unfurlUrlToActionApiUrl = _chunkXNEBNDEXcjs.unfurlUrlToBlinkApiUrl; exports.unfurlUrlToBlinkApiUrl = _chunkXNEBNDEXcjs.unfurlUrlToBlinkApiUrl; exports.useAction = _chunkROVRYL42cjs.useAction; exports.useActionApiUrl = _chunkROVRYL42cjs.useBlinkApiUrl; exports.useActionsRegistryInterval = _chunkROVRYL42cjs.useBlinksRegistryInterval; exports.useBlink = _chunkROVRYL42cjs.useBlink2; exports.useBlinkApiUrl = _chunkROVRYL42cjs.useBlinkApiUrl; exports.useBlinkList = _chunkROVRYL42cjs.useBlinkList2; exports.useBlinksRegistryInterval = _chunkROVRYL42cjs.useBlinksRegistryInterval; exports.useMetadata = _chunkROVRYL42cjs.useMetadata; exports.verifySignMessageData = _chunkXNEBNDEXcjs.verifySignMessageData;
