// @dialectlabs/blinks-core/dist/solana/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class;

var _chunkXNEBNDEXcjs = require('../chunk-XNEBNDEX.cjs');

// src/solana/BlinkSolanaConfig.ts
var _web3js = require('@solana/web3.js');
var BlinkSolanaConfig = (_class = class _BlinkSolanaConfig {
  constructor(rpcUrlOrConnection, adapter) {
    this.adapter = adapter;
    if (!rpcUrlOrConnection) {
      throw new Error("rpcUrl or connection is required");
    }
    this.connection = typeof rpcUrlOrConnection === "string" ? new (0, _web3js.Connection)(rpcUrlOrConnection, "confirmed") : rpcUrlOrConnection;
  }
  static __initStatic() {this.CONFIRM_TIMEOUT_MS = 6e4 * 1.2}
  // 20% extra time
  static __initStatic2() {this.DEFAULT_METADATA = {
    supportedBlockchainIds: _chunkXNEBNDEXcjs.DEFAULT_SUPPORTED_BLOCKCHAIN_IDS
  }}
  
  get metadata() {
    return _nullishCoalesce(this.adapter.metadata, () => ( _BlinkSolanaConfig.DEFAULT_METADATA));
  }
  signTransaction(tx, context) {
    return this.adapter.signTransaction(tx, context);
  }
  confirmTransaction(signature) {
    return new Promise((res, rej) => {
      const start = Date.now();
      const confirm = async () => {
        if (Date.now() - start >= _BlinkSolanaConfig.CONFIRM_TIMEOUT_MS) {
          rej(new Error("Unable to confirm transaction: timeout reached"));
          return;
        }
        try {
          const status = await this.connection.getSignatureStatus(signature);
          if (_optionalChain([status, 'access', _ => _.value, 'optionalAccess', _2 => _2.err])) {
            rej(
              new Error(
                `Transaction execution failed, check wallet for details`
              )
            );
            return;
          }
          if (status.value && status.value.confirmations !== null) {
            res();
            return;
          }
        } catch (e) {
          console.error(
            "[@dialectlabs/blinks-core] Error confirming transaction",
            e
          );
          rej(
            new Error(
              "Error confirming transaction. Please check your RPC connection and transaction status"
            )
          );
          return;
        }
        setTimeout(confirm, 3e3);
      };
      confirm();
    });
  }
  async signMessage(data, context) {
    return this.adapter.signMessage(data, context);
  }
  async connect(context) {
    try {
      return await this.adapter.connect(context);
    } catch (e2) {
      return null;
    }
  }
}, _class.__initStatic(), _class.__initStatic2(), _class);



exports.ActionConfig = BlinkSolanaConfig; exports.BlinkSolanaConfig = BlinkSolanaConfig;
